rclcpp/test/rclcpp/executors/test_multi_threaded_executor.cpp

```diff
@@ -102,3 +102,66 @@ TEST_F(TestMultiThreadedExecutor, timer_over_take) {
   executor.add_node(node);
   executor.spin();
 }
+
+/*
+   Test that no tasks are starved
+ */
+TEST_F(TestMultiThreadedExecutor, starvation) {
```

Comment:
why this should be assertion for the test? if we create the custom callback group with `rclcpp::CallbackGroupType::Reentrant` and add it to `create_wall_timer`, this test should pass because there are 2 threads that executor can assign the executables concurrently.
in other word, this is expected behavior that user specifies with `MutuallyExclusive` (default) as we discussed on https://github.com/ros2/rclcpp/issues/2645?

`SingleThreadedExecutor` has the exact same situation like this. if you are trying to fix this timer starvation in the `Executor`, i do not think that is the problem only for `MultiThreadedExecutor`, if the timer callback overruns.

> The executor should be alternating between these two tasks, never executing one task twice before the other.

IMO, i am not sure if this assumption is correct by system. sounds like user requirement, user would want to have a high priority timer which they want it to be executed as fast as possible. in that case, this fix becomes the problem for that requirement. (in that case, what's missing here is priority order user interface?)

CC: @jmachowinski @mjcarroll @alsora

Reply:
In the unit-test you are creating two timers with a period of `0ms`.
In this scenario there's no guarantee that both timers will be called, because they are both always ready all the time. Note that the problem is not that they have the same period, but rather that the period is `0ms`.
Since they are both always ready, only the "first one" will be invoked.

Can you please check what happens if the period is set to `10ms` for both?